src/App.tsx

```diff
@@ -0,0 +1,795 @@
+import React, { ReactElement, useCallback, useEffect, useMemo, useState } from "react";
+import { colorRamps, Plotting, ColorizeCanvas, Dataset, Track } from "./colorizer";
+import { BACKGROUND_ID } from "./colorizer/ColorizeCanvas";
+import RecordingControls from "./colorizer/RecordingControls";
+import TimeControls from "./colorizer/TimeControls";
+import * as urlUtils from "./colorizer/utils/url_utils";
+
+import styles from "./App.module.css";
+import { useDebounce } from "./colorizer/utils/react_utils";
+
+const CANVAS_PLACEHOLDER_ID = "canvasPlaceholder";
+const COLOR_RAMP_PLACEHOLDER_ID = "colorRamp";
+const PLOT_PLACEHOLDER_ID = "plot";
+const DEFAULT_COLOR_RAMP = 4;
+
+function App(): ReactElement {
+  // STATE INITIALIZATION /////////////////////////////////////////////////////////
+
+  const [plot, setPlot] = useState<Plotting | null>(null);
+  const canv = useMemo(() => {
```

Comment:
I know we talked about this but I am still trying to understand.  It seems this is more like a single-shot useEffect because it is creating the ColorizeCanvas, than a useMemo because it's not computing something that might ever need to be re-computed.  In this case maybe useEffect and useMemo have no real difference.

Reply:
There's a good quick description of the difference between the hooks here: https://www.geeksforgeeks.org/when-to-use-usecallback-usememo-and-useeffect

Quick paraphrasing:
`useMemo` returns the memoized **value** of a function that is provided, while `useEffect` is a callback run **after all the components have been rendered** and **does not return a value.**

While both only recalculate the provided function when their dependencies change, they happen in different parts of the React lifecycle (immediately for `useMemo` vs. after the render for `useEffect`), and only one (`useMemo`) returns its results. 

Other things I could have used:
```
// This would have also been fine, but because I never set the canvas again it adds an unused variable.
const [canv, setCanv] = useState(new ColorizeCanvas());

// useEffect only makes sense if used with state, since it doesn't return any values.
// This would also lead to some weirdness with canv being undefined during the first render.
const [canv, setCanv] = useState<ColorizeCanvas>();
useEffect(() => {
  setCanv(new ColorizeCanvas());
}, []);
```

Reply:
Thank you!  That does help and I forgot about needing the return value.